mypyc/ir/rtypes.py

```diff
@@ -436,6 +446,131 @@ def deserialize(cls, data: JsonDict, ctx: 'DeserMaps') -> 'RTuple':
+    """
+    unaligned_sizes = [compute_rtype_size(typ) for typ in types]
+    alignments = [compute_rtype_alignment(typ) for typ in types]
+
+    current_offset = 0
+    offsets = []
+    final_size = 0
+    for i in range(len(unaligned_sizes)):
+        offsets.append(current_offset)
+        if i + 1 < len(unaligned_sizes):
+            cur_size = unaligned_sizes[i]
+            current_offset += cur_size
+            next_alignment = alignments[i + 1]
+            # compute aligned offset,
+            # check https://en.wikipedia.org/wiki/Data_structure_alignment for more infomation
+            current_offset = (current_offset + (next_alignment - 1)) & -next_alignment
+        else:
+            struct_alignment = max(alignments)
+            final_size = current_offset + unaligned_sizes[i]
+            final_size = (final_size + (struct_alignment - 1)) & -struct_alignment
+    return offsets, final_size
+
+
+class StructInfo:
+    """Struct-like type Infomation
+
+    StructInfo should work with registry to ensure constraints like the unique naming
+    constraint for struct type
+    """
+    def __init__(self,
+                 name: str,
+                 names: List[str],
+                 types: List[RType]) -> None:
+        self.name = name
+        self.names = names
+        self.types = types
+        # generate dummy names
+        if len(self.names) < len(self.types):
+            for i in range(len(self.types) - len(self.names)):
+                self.names.append('_item' + str(i))
+        self.offsets, self.size = compute_aligned_offsets_and_size(types)
+
+
+class RStruct(RType):
+    """Represent CPython structs"""
+    def __init__(self,
+                 info: StructInfo) -> None:
+        self.name = info.name
+        self.names = info.names
+        self.types = info.types
+        self.offsets = info.offsets
+        self.size = info.size
+        self._ctype = info.name
+
+    def accept(self, visitor: 'RTypeVisitor[T]') -> T:
+        return visitor.visit_rstruct(self)
+
+    def __str__(self) -> str:
+        # if not tuple(unamed structs)
+        return '%s{%s}' % (self.name, ', '.join(name + ":" + str(typ)
+                                                for name, typ in zip(self.names, self.types)))
+
+    def __repr__(self) -> str:
+        return '<RStruct %s{%s}>' % (self.name, ', '.join(name + ":" + repr(typ) for name, typ
+                                                          in zip(self.names, self.types)))
+
+    def __eq__(self, other: object) -> bool:
+        return (isinstance(other, RStruct) and self.name == other.name
+                and self.names == other.names and self.types == other.types)
+
+    def __hash__(self) -> int:
+        return hash((self.name, self.names, self.types))
+
+    def serialize(self) -> JsonDict:
+        types = [x.serialize() for x in self.types]
+        return {'.class': 'RStruct', 'name': self.name, 'names': self.names, 'types': types}
```

Comment:
With the above suggestion it's sufficient to serialize the name of the struct. However, this means that we'll need to serialize the struct registry somewhere else. It's okay to just `assert False` here for now, since we won't need to serialize struct types yet, I think.